File: 3.2D_Arrays/java/MatrixAddition.java
/**
 * Matrix Addition
 * 
 * This program demonstrates how to add two matrices in Java.
 * 
 * Time Complexity: O(m×n) where m and n are the dimensions of the matrices
 * Space Complexity: O(m×n) for storing the result matrix
 */

public class MatrixAddition {
    
    /**
     * Add two matrices and return the result.
     * 
     * @param matrix1 The first input matrix
     * @param matrix2 The second input matrix
     * @return A new matrix containing the sum of the input matrices
     * @throws IllegalArgumentException if matrices have different dimensions
     */
    public static int[][] addMatrices(int[][] matrix1, int[][] matrix2) {
        // Check if matrices are null
        if (matrix1 == null || matrix2 == null) {
            throw new IllegalArgumentException("Matrices cannot be null");
        }
        
        // Check if matrices are empty
        if (matrix1.length == 0 || matrix2.length == 0) {
            throw new IllegalArgumentException("Matrices cannot be empty");
        }
        
        // Check if matrices have the same dimensions
        if (matrix1.length != matrix2.length || matrix1[0].length != matrix2[0].length) {
            throw new IllegalArgumentException("Matrices must have the same dimensions for addition");
        }
        
        int rows = matrix1.length;
        int cols = matrix1[0].length;
        
        // Initialize result matrix
        int[][] result = new int[rows][cols];
        
        // Add corresponding elements
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                result[i][j] = matrix1[i][j] + matrix2[i][j];
            }
        }
        
        return result;
    }
    
    /**
     * Print a matrix in a readable format.
     * 
     * @param matrix The matrix to print
     */
    public static void printMatrix(int[][] matrix) {
        for (int[] row : matrix) {
            for (int val : row) {
                System.out.print(val + "\t");
            }
            System.out.println();
        }
    }
    
    public static void main(String[] args) {
        // Example matrices
        int[][] matrix1 = {
            {1, 2, 3},
            {4, 5, 6},
            {7, 8, 9}
        };
        
        int[][] matrix2 = {
            {9, 8, 7},
            {6, 5, 4},
            {3, 2, 1}
        };
        
        System.out.println("Matrix 1:");
        printMatrix(matrix1);
        
        System.out.println("\nMatrix 2:");
        printMatrix(matrix2);
        
        try {
            // Add the matrices
            int[][] result = addMatrices(matrix1, matrix2);
            
            System.out.println("\nResult of Matrix Addition:");
            printMatrix(result);
        } catch (IllegalArgumentException e) {
            System.err.println("Error: " + e.getMessage());
        }
        
        // Example with matrices of different dimensions
        System.out.println("\nExample with matrices of different dimensions:");
        int[][] matrix3 = {
            {1, 2},
            {3, 4}
        };
        
        System.out.println("Matrix 3 (2x2):");
        printMatrix(matrix3);
        
        try {
            // Try to add matrices of different dimensions
            int[][] result = addMatrices(matrix1, matrix3);
            printMatrix(result);
        } catch (IllegalArgumentException e) {
            System.err.println("Error: " + e.getMessage());
        }
    }
}

File: 3.2D_Arrays/Kotlin/MatrixAddition.kt
/**
 * Matrix Addition
 * 
 * This program demonstrates how to add two matrices in Kotlin using both
 * imperative and functional approaches.
 * 
 * Time Complexity: O(m×n) where m and n are the dimensions of the matrices
 * Space Complexity: O(m×n) for storing the result matrix
 */

/**
 * Add two matrices using an imperative approach.
 * 
 * @param matrix1 The first input matrix
 * @param matrix2 The second input matrix
 * @return A new matrix containing the sum of the input matrices
 * @throws IllegalArgumentException if matrices have different dimensions
 */
fun addMatrices(matrix1: Array<IntArray>, matrix2: Array<IntArray>): Array<IntArray> {
    // Check if matrices have the same dimensions
    if (matrix1.isEmpty() || matrix2.isEmpty()) {
        throw IllegalArgumentException("Matrices cannot be empty")
    }
    
    if (matrix1.size != matrix2.size || matrix1[0].size != matrix2[0].size) {
        throw IllegalArgumentException("Matrices must have the same dimensions for addition")
    }
    
    val rows = matrix1.size
    val cols = matrix1[0].size
    
    // Initialize result matrix
    val result = Array(rows) { IntArray(cols) }
    
    // Add corresponding elements
    for (i in 0 until rows) {
        for (j in 0 until cols) {
            result[i][j] = matrix1[i][j] + matrix2[i][j]
        }
    }
    
    return result
}

/**
 * Add two matrices using a functional approach.
 * 
 * @param matrix1 The first input matrix
 * @param matrix2 The second input matrix
 * @return A new matrix containing the sum of the input matrices
 * @throws IllegalArgumentException if matrices have different dimensions
 */
fun addMatricesFunctional(matrix1: Array<IntArray>, matrix2: Array<IntArray>): Array<IntArray> {
    // Check if matrices have the same dimensions
    if (matrix1.isEmpty() || matrix2.isEmpty()) {
        throw IllegalArgumentException("Matrices cannot be empty")
    }
    
    if (matrix1.size != matrix2.size || matrix1[0].size != matrix2[0].size) {
        throw IllegalArgumentException("Matrices must have the same dimensions for addition")
    }
    
    // Use functional approach to create the result matrix
    return Array(matrix1.size) { i ->
        IntArray(matrix1[0].size) { j ->
            matrix1[i][j] + matrix2[i][j]
        }
    }
}

/**
 * Print a matrix in a readable format.
 * 
 * @param matrix The matrix to print
 */
fun printMatrix(matrix: Array<IntArray>) {
    matrix.forEach { row ->
        row.forEach { value ->
            print("$value\t")
        }
        println()
    }
}

fun main() {
    // Example matrices
    val matrix1 = arrayOf(
        intArrayOf(1, 2, 3),
        intArrayOf(4, 5, 6),
        intArrayOf(7, 8, 9)
    )
    
    val matrix2 = arrayOf(
        intArrayOf(9, 8, 7),
        intArrayOf(6, 5, 4),
        intArrayOf(3, 2, 1)
    )
    
    println("Matrix 1:")
    printMatrix(matrix1)
    
    println("\nMatrix 2:")
    printMatrix(matrix2)
    
    try {
        // Add the matrices using imperative approach
        val result = addMatrices(matrix1, matrix2)
        
        println("\nResult of Matrix Addition (Imperative Approach):")
        printMatrix(result)
        
        // Add the matrices using functional approach
        val resultFunctional = addMatricesFunctional(matrix1, matrix2)
        
        println("\nResult of Matrix Addition (Functional Approach):")
        printMatrix(resultFunctional)
    } catch (e: IllegalArgumentException) {
        println("Error: ${e.message}")
    }
    
    // Example with matrices of different dimensions
    println("\nExample with matrices of different dimensions:")
    val matrix3 = arrayOf(
        intArrayOf(1, 2),
        intArrayOf(3, 4)
    )
    
    println("Matrix 3 (2x2):")
    printMatrix(matrix3)
    
    try {
        // Try to add matrices of different dimensions
        val result = addMatrices(matrix1, matrix3)
        printMatrix(result)
    } catch (e: IllegalArgumentException) {
        println("Error: ${e.message}")
    }
}